dpnp/backend/extensions/ufunc/elementwise_functions/nan_to_num.cpp

```diff
@@ -0,0 +1,425 @@
+{
+    sycl::event to_num_ev;
+
+    using dpctl::tensor::type_utils::is_complex;
+    if constexpr (is_complex<T>::value) {
+        using realT = typename T::value_type;
+        realT nan_v = py::cast<realT>(py_nan);
+        realT posinf_v = py::cast<realT>(py_posinf);
+        realT neginf_v = py::cast<realT>(py_neginf);
+
+        using dpnp::kernels::nan_to_num::nan_to_num_impl;
+        to_num_ev = nan_to_num_impl<T, realT>(
+            exec_q, nd, nelems, shape_strides, nan_v, posinf_v, neginf_v, arg_p,
+            arg_offset, dst_p, dst_offset, depends);
+    }
+    else {
+        T nan_v = py::cast<T>(py_nan);
+        T posinf_v = py::cast<T>(py_posinf);
+        T neginf_v = py::cast<T>(py_neginf);
+
+        using dpnp::kernels::nan_to_num::nan_to_num_impl;
+        to_num_ev = nan_to_num_impl<T, T>(
+            exec_q, nd, nelems, shape_strides, nan_v, posinf_v, neginf_v, arg_p,
+            arg_offset, dst_p, dst_offset, depends);
+    }
+    return to_num_ev;
+}
+
+typedef sycl::event (*nan_to_num_contig_fn_ptr_t)(
+    sycl::queue &,
+    size_t,
+    const py::object &,
+    const py::object &,
+    const py::object &,
+    const char *,
+    char *,
+    const std::vector<sycl::event> &);
+
+template <typename T>
+sycl::event nan_to_num_contig_call(sycl::queue &exec_q,
+                                   size_t nelems,
+                                   const py::object &py_nan,
+                                   const py::object &py_posinf,
+                                   const py::object &py_neginf,
+                                   const char *arg_p,
+                                   char *dst_p,
+                                   const std::vector<sycl::event> &depends)
+{
+    sycl::event to_num_contig_ev;
+
+    using dpctl::tensor::type_utils::is_complex;
+    if constexpr (is_complex<T>::value) {
+        using realT = typename T::value_type;
+        realT nan_v = py::cast<realT>(py_nan);
+        realT posinf_v = py::cast<realT>(py_posinf);
+        realT neginf_v = py::cast<realT>(py_neginf);
+
+        using dpnp::kernels::nan_to_num::nan_to_num_contig_impl;
+        to_num_contig_ev = nan_to_num_contig_impl<T, realT>(
+            exec_q, nelems, nan_v, posinf_v, neginf_v, arg_p, dst_p, depends);
+    }
+    else {
+        T nan_v = py::cast<T>(py_nan);
+        T posinf_v = py::cast<T>(py_posinf);
+        T neginf_v = py::cast<T>(py_neginf);
+
+        using dpnp::kernels::nan_to_num::nan_to_num_contig_impl;
+        to_num_contig_ev = nan_to_num_contig_impl<T, T>(
+            exec_q, nelems, nan_v, posinf_v, neginf_v, arg_p, dst_p, depends);
+    }
+    return to_num_contig_ev;
+}
+
+namespace td_ns = dpctl::tensor::type_dispatch;
+nan_to_num_fn_ptr_t nan_to_num_dispatch_vector[td_ns::num_types];
+nan_to_num_contig_fn_ptr_t nan_to_num_contig_dispatch_vector[td_ns::num_types];
+
+std::pair<sycl::event, sycl::event>
+    py_nan_to_num(const dpctl::tensor::usm_ndarray &src,
```

Comment:
Yes, we can probably separate some out—and even when py_unary_ufunc and py_binary_ufunc are finally exposed, it will still make sense to have for eventual inclusion of `divmod`, `frexp`, and `modf`.